shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1832,6 +1903,85 @@ - (void)scrollEvent:(UIPanGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4))
     _mouseState.last_translation = CGPointZero;
   }
 
+  if (_mouseState.hover_pointer_is_added) {
+    packet->SetPointerData(/*index=*/0, pointer_data);
+  } else {
+    flutter::PointerData add_pointer_data = pointer_data;
+    add_pointer_data.signal_kind = flutter::PointerData::SignalKind::kNone;
+    add_pointer_data.change = flutter::PointerData::Change::kAdd;
+    _mouseState.hover_pointer_is_added = true;
+    packet->SetPointerData(/*index=*/0, add_pointer_data);
+    packet->SetPointerData(/*index=*/1, pointer_data);
+  }
+
+  [_engine.get() dispatchPointerDataPacket:std::move(packet)];
+}
+
+- (void)panEvent:(UIPanGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4)) {
+  CGPoint translation = [recognizer translationInView:self.view];
+  const CGFloat scale = [UIScreen mainScreen].scale;
+
+  translation.x *= scale;
+  translation.y *= scale;
+
+  flutter::PointerData pointer_data =
+      [self generateDataForHoverPointerWithDeviceKind:flutter::PointerData::DeviceKind::kTrackpad];
+  // Ensure that different touches have different devices.
+  pointer_data.device = reinterpret_cast<int64_t>(recognizer);
+  switch (recognizer.state) {
+    case UIGestureRecognizerStateBegan:
+      pointer_data.change = flutter::PointerData::Change::kPanZoomStart;
+      break;
+    case UIGestureRecognizerStateChanged:
+      pointer_data.change = flutter::PointerData::Change::kPanZoomUpdate;
+      pointer_data.pan_x = translation.x;
+      pointer_data.pan_y = translation.y;
+      pointer_data.pan_delta_x = 0;  // Delta will be generated in pointer_data_packet_converter.cc.
+      pointer_data.pan_delta_y = 0;  // Delta will be generated in pointer_data_packet_converter.cc.
+      pointer_data.scale = 1;
+      break;
+    case UIGestureRecognizerStateEnded:
+    case UIGestureRecognizerStateCancelled:
+      pointer_data.change = flutter::PointerData::Change::kPanZoomEnd;
+      break;
+    default:
+      // panEvent: should only ever be triggered with the above phases
+      NSAssert(false, @"Trackpad pan event occured with unexpected phase 0x%lx",
+               (long)recognizer.state);
+      break;
+  }
+
+  auto packet = std::make_unique<flutter::PointerDataPacket>(1);
+  packet->SetPointerData(/*index=*/0, pointer_data);
+
+  [_engine.get() dispatchPointerDataPacket:std::move(packet)];
+}
+
+- (void)pinchEvent:(UIPinchGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4)) {
+  auto packet = std::make_unique<flutter::PointerDataPacket>(1);
+
+  flutter::PointerData pointer_data =
+      [self generateDataForHoverPointerWithDeviceKind:flutter::PointerData::DeviceKind::kMouse];
```

Comment:
should this be trackpad instead of kMouse? 

also same question - why are we sending hover pointer in pinch callback? (instead of in a hover recognizer's callback.

Reply:
Laziness / reuse of similar code. It was not as obvious before the name of `generateDataForHoverPointerWithDeviceKind` was changed. I will fix it.